### ISSUE TITLE
Generated Terraform code related to vpc_ipv6_cidr_block fails

### ISSUE THREAD

**Issue by felixfriedrich** [kubernetes/kops#15245]
/kind bug

**1. What `kops` version are you running? The command `kops version`, will display
 this information.**

`v1.26.2`

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**

`v1.25.2`

**3. What cloud provider are you using?**

AWS

**4. What commands did you run?  What is the simplest way to reproduce this issue?**

We generate Terraform code based on a `cluster.yaml`:
```
apiVersion: kops.k8s.io/v1alpha2
kind: Cluster
...
```

* `kops replace -f cluster.yml`
* `kops update cluster --target=terraform --name $(CLUSTER_NAME) --out=.`

The two commands above generate Terraform code which contains the following bits, which are for the sake of this bug report as minimal as possible:

```
data "aws_vpc" "my_vpc" {
  id = "vpc-REDACTED"
}

locals {
  vpc_ipv6_cidr_block  = data.aws_vpc.my_vpc.ipv6_cidr_block
  vpc_ipv6_cidr_length = local.vpc_ipv6_cidr_block == null ? null : tonumber(regex(".*/(\\d+)", local.vpc_ipv6_cidr_block)[0])
}
```

**5. What happened after the commands executed?**

This code fails whoever with:
```
│ Error: Error in function call
│
│   on kops-bug.tf line 7, in locals:
│    7:   vpc_ipv6_cidr_length = local.vpc_ipv6_cidr_block == null ? null : tonumber(regex(".*/(\\d+)", local.vpc_ipv6_cidr_block)[0])
│     ├────────────────
│     │ while calling regex(pattern, string)
│     │ local.vpc_ipv6_cidr_block is ""
│
│ Call to function "regex" failed: pattern did not match any part of the given string.
```

**6. What did you expect to happen?**

Correcting the code like below works:
```
local.vpc_ipv6_cidr_block == null ? null : tonumber(regex(".*/(\\d+)", local.vpc_ipv6_cidr_block)[0])
```

**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -o yaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.**

```yaml
apiVersion: kops.k8s.io/v1alpha2
kind: Cluster
metadata:
  creationTimestamp: null
  name: cluster-1.example.com
spec:
  api:
    loadBalancer:
      class: Network
      type: Public
  authorization:
    rbac: {}
  channel: stable
  cloudLabels:
    project: shared
    stage: production
    team: A
  cloudProvider: aws
  configBase: s3://bucket-name/production/k8s/cluster-1.example.com
  configStore: s3://bucket-name/production/k8s/cluster-1.example.com
  containerRuntime: containerd
  dnsZone: example.com
  etcdClusters:
  - etcdMembers:
    - encryptedVolume: true
      instanceGroup: master-a
      name: a
    - encryptedVolume: true
      instanceGroup: master-b
      name: b
    - encryptedVolume: true
      instanceGroup: master-c
      name: c
    manager:
      env:
      - name: ETCD_MANAGER_HOURLY_BACKUPS_RETENTION
        value: 1d
      - name: ETCD_MANAGER_DAILY_BACKUPS_RETENTION
        value: 7d
    name: main
  - etcdMembers:
    - encryptedVolume: true
      instanceGroup: master-a
      name: a
    - encryptedVolume: true
      instanceGroup: master-b
      name: b
    - encryptedVolume: true
      instanceGroup: master-c
      name: c
    manager:
      env:
      - name: ETCD_MANAGER_HOURLY_BACKUPS_RETENTION
        value: 1d
      - name: ETCD_MANAGER_DAILY_BACKUPS_RETENTION
        value: 7d
    name: events
  - etcdMembers:
    - encryptedVolume: true
      instanceGroup: master-a
      name: a
    - encryptedVolume: true
      instanceGroup: master-b
      name: b
    - encryptedVolume: true
      instanceGroup: master-c
      name: c
    manager:
      env:
      - name: ETCD_AUTO_COMPACTION_MODE
        value: revision
      - name: ETCD_AUTO_COMPACTION_RETENTION
        value: "2500"
    name: cilium
  iam:
    allowContainerRegistry: true
    legacy: false
  kubeAPIServer:
    logLevel: 1
    oidcClientID: REDACTED.apps.googleusercontent.com
    oidcIssuerURL: https://accounts.google.com
    oidcUsernameClaim: email
  kubeControllerManager:
    horizontalPodAutoscalerSyncPeriod: 15s
    logLevel: 1
  kubeDNS:
    provider: CoreDNS
  kubeProxy:
    logLevel: 1
  kubeScheduler:
    logLevel: 1
  kubelet:
    anonymousAuth: false
    authenticationTokenWebhook: true
    authorizationMode: Webhook
    logLevel: 1
  kubernetesApiAccess:
  - 0.0.0.0/0
  kubernetesVersion: 1.25.2
  masterPublicName: api.cluster-1.example.com
  networkCIDR: REDACTED.REDACTED.REDACTED.REDACTED/16
  networkID: vpc-REDACTED
  networking:
    cilium:
      etcdManaged: true
  nonMasqueradeCIDR: REDACTED.REDACTED.REDACTED.REDACTED/10
  sshAccess:
  - 0.0.0.0/0
  sshKeyName: aws_cluster_1_production
  subnets:
  - cidr: REDACTED.REDACTED.REDACTED.REDACTED/26
    name: utility-a
    type: Utility
    zone: us-east-1a
  - cidr: REDACTED.REDACTED.REDACTED.REDACTED/26
    name: utility-b
    type: Utility
    zone: us-east-1b
  - cidr: REDACTED.REDACTED.REDACTED.REDACTED/26
    name: utility-c
    type: Utility
    zone: us-east-1c
  - cidr: REDACTED.REDACTED.REDACTED.REDACTED/24
    name: a
    type: Private
    zone: us-east-1a
  - cidr: REDACTED.REDACTED.REDACTED.REDACTED/24
    name: b
    type: Private
    zone: us-east-1b
  - cidr: REDACTED.REDACTED.REDACTED.REDACTED/24
    name: c
    type: Private
    zone: us-east-1c
  topology:
    bastion:
      bastionPublicName: bastion.cluster-1.example.com
    dns:
      type: Public
    masters: private
    nodes: private

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2023-03-17T14:03:01Z"
  labels:
    kops.k8s.io/cluster: cluster-1.example.com
  name: bastions
spec:
  image: REDACTED/ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20220924
  machineType: t2.micro
  maxSize: 1
  minSize: 1
  role: Bastion
  subnets:
  - utility-a
  - utility-b
  - utility-c

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2023-03-17T14:03:03Z"
  labels:
    kops.k8s.io/cluster: cluster-1.example.com
  name: master-a
spec:
  image: 099720109477/ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20220924
  machineType: c5.xlarge
  maxSize: 1
  minSize: 1
  role: Master
  subnets:
  - a

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2023-03-17T14:03:05Z"
  labels:
    kops.k8s.io/cluster: cluster-1.example.com
  name: master-b
spec:
  image: 099720109477/ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20220924
  machineType: c5.xlarge
  maxSize: 1
  minSize: 1
  role: Master
  subnets:
  - b

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2023-03-17T14:03:07Z"
  labels:
    kops.k8s.io/cluster: cluster-1.example.com
  name: master-c
spec:
  image: 099720109477/ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20220924
  machineType: c5.xlarge
  maxSize: 1
  minSize: 1
  role: Master
  subnets:
  - c

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2023-03-17T14:03:09Z"
  labels:
    kops.k8s.io/cluster: cluster-1.example.com
  name: nodes-a
spec:
  additionalSecurityGroups:
  - sg-REDACTED
  cloudLabels:
    k8s.io/cluster-autoscaler/enabled: "1"
    k8s.io/cluster-autoscaler/cluster-1.example.com: "1"
  image: 099720109477/ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20220924
  machineType: c5a.2xlarge
  maxSize: 250
  minSize: 0
  role: Node
  subnets:
  - a
  warmPool:
    maxSize: 5
    minSize: 3

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2023-03-17T14:03:10Z"
  labels:
    kops.k8s.io/cluster: cluster-1.example.com
  name: nodes-b
spec:
  additionalSecurityGroups:
  - sg-REDACTED
  cloudLabels:
    k8s.io/cluster-autoscaler/enabled: "1"
    k8s.io/cluster-autoscaler/cluster-1.example.com: "1"
  image: 099720109477/ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20220924
  machineType: c5a.2xlarge
  maxSize: 250
  minSize: 0
  role: Node
  subnets:
  - b
  warmPool:
    maxSize: 5
    minSize: 3

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2023-03-17T14:03:12Z"
  labels:
    kops.k8s.io/cluster: cluster-1.example.com
  name: nodes-c
spec:
  additionalSecurityGroups:
  - sg-REDACTED
  cloudLabels:
    k8s.io/cluster-autoscaler/enabled: "1"
    k8s.io/cluster-autoscaler/cluster-1.example.com: "1"
  image: 099720109477/ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20220924
  machineType: c5a.2xlarge
  maxSize: 250
  minSize: 0
  role: Node
  subnets:
  - c
  warmPool:
    maxSize: 5
    minSize: 3

```

**8. Please run the commands with most verbose logging by adding the `-v 10` flag.
  Paste the logs into this report, or in a gist and provide the gist link here.**

**9. Anything else do we need to know?**

**Comment by hakman** [comment]
@felixfriedrich What Terraform version are you using?

**Comment by felixfriedrich** [comment]
@hakman We're using Terraform version `1.3.7`.

**Comment by felixfriedrich** [comment]
> @felixfriedrich That's true, tested it today. The correct approach would be to compare with `""` instead of `null`, or not add this at all to the generated file:
> 
> ```terraform
> vpc_ipv6_cidr_length = local.vpc_ipv6_cidr_block == "" ? null : tonumber(regex(".*/(\\d+)", local.vpc_ipv6_cidr_block)[0])
> ```

Would you accept a PR to fix this? Or would you rather fix this yourself?

**Comment by hakman** [comment]
> Would you accept a PR to fix this? Or would you rather fix this yourself?

I think fix is https://github.com/kubernetes/kops/pull/15262. Let's see if that merges. If not, a PR would be welcomed.

**Comment by felixfriedrich** [comment]
@hakman I used your branch to see whether the behaviour changes. It doesn't unfortunately. I am happy to share redacted details about our VPC, but the gist would be that there is nothing ipv6 related. We use this Terraform module to create the VPC: https://github.com/terraform-aws-modules/terraform-aws-vpc

And the module call is pretty much like this:

```
module "vpc" {
  source = "github.com/terraform-aws-modules/terraform-aws-vpc"

  name = "onrunning"
  cidr = "X.X.0.0/16"

  azs = [
    "us-east-1a",
    "us-east-1b",
    "us-east-1c",
    "us-east-1d"
  ]

  private_subnets = [
    "X.X.X.X/21",
    "X.X.X.X/21",
    ...    
  ]

  public_subnets = [
    "X.X.X.X/21",
    "X.X.X.X/21",
    ...    
  ]

  enable_nat_gateway      = true
  single_nat_gateway      = true
  one_nat_gateway_per_az  = false
  map_public_ip_on_launch = false
  enable_dns_support      = true
  enable_dns_hostnames    = true
}
```

**Comment by hakman** [comment]
@felixfriedrich Would you like to try updating the compare function with `""` instead of `null`?
https://github.com/kubernetes/kops/pull/15262#issuecomment-1484481158

**Comment by felixfriedrich** [comment]
Yes, absolutely.

**Comment by dirtyValera** [comment]
Is this fixed? Recently installed kops 1.26.2 and am getting this error. What do I do?
